cli/cook/cli.py

```diff
@@ -39,7 +40,22 @@ def run(args, plugins):
     Main entrypoint to the cook scheduler CLI. Loads configuration files, 
     processes global command line arguments, and calls other command line 
     sub-commands (actions) if necessary.
+
+    plugins is a map from plugin-name -> function or Class.SubCommandPlugin
     """
+
+    # This has to happen before we parse the args, otherwise we might
+    # get subcommand not found.
+    for name, instance in plugins.items():
+        if isinstance(instance, SubCommandPlugin):
+            logging.debug('Adding SubCommandPlugin %s' % name)
+            instance.register(subparsers.add_parser, configuration.add_defaults)
+            logging.debug('Done adding SubCommandPlugin %s' % name)
+            name = instance.name()
+            if name in actions:
+                raise Exception('SubCommandPlugin %s clashes with an existing subcommand.' % name)
+            actions[name] = instance.run
```

Comment:
Maybe we should wrap these lines in a `try`/`catch` so that one bad plugin doesn't blow everything up?

Reply:
yeah let me do that.